src/FindStmtCost.cpp

```diff
@@ -0,0 +1,930 @@
+    };
+
+    std::function<int(int)> calculate_dmc = [scaling_factor](int children_cost) {
+        return scaling_factor * (LOAD_DM_COST + children_cost);
+    };
+
+    // inclusive and exclusive costs are the same
+    set_costs(true, op, {op->predicate.get(), op->index.get()}, calculate_cc, calculate_dmc);
+    set_costs(false, op, {op->predicate.get(), op->index.get()}, calculate_cc, calculate_dmc);
+}
+
+void FindStmtCost::visit(const Ramp *op) {
+    op->base.accept(this);
+    op->stride.accept(this);
+
+    // inclusive and exclusive costs are the same
+    set_costs(true, op, {op->base.get(), op->stride.get()});
+    set_costs(false, op, {op->base.get(), op->stride.get()});
+}
+
+void FindStmtCost::visit(const Broadcast *op) {
+    op->value.accept(this);
+
+    // inclusive and exclusive costs are the same
+    set_costs(true, op, {op->value.get()});
+    set_costs(false, op, {op->value.get()});
+}
+
+void FindStmtCost::visit(const Call *op) {
+    vector<const IRNode *> children;
+
+    for (const auto &arg : op->args) {
+        arg.accept(this);
+        children.push_back(arg.get());
+    }
+
+    // Consider extern call args
+    if (op->func.defined()) {
+        Function f(op->func);
+        if (op->call_type == Call::Halide && f.has_extern_definition()) {
+            for (const auto &arg : f.extern_arguments()) {
+                if (arg.is_expr()) {
+                    arg.expr.accept(this);
+                    children.push_back(arg.expr.get());
+                }
+            }
+        }
+    }
+
+    // inclusive and exclusive costs are the same
+    set_costs(true, op, children);
+    set_costs(false, op, children);
+}
+
+void FindStmtCost::visit(const Let *op) {
+
+    op->value.accept(this);
+    op->body.accept(this);
+
+    // inclusive and exclusive costs are the same (keep body in both since it's all inlined)
+    set_costs(true, op, {op->value.get(), op->body.get()});
+    set_costs(false, op, {op->value.get(), op->body.get()});
+}
+
+void FindStmtCost::visit(const Shuffle *op) {
+    vector<const IRNode *> children;
+    for (const Expr &i : op->vectors) {
+        i.accept(this);
+        children.push_back(i.get());
+    }
+
+    // inclusive and exclusive costs are the same
+    set_costs(true, op, children);
+    set_costs(false, op, children);
+}
+
+void FindStmtCost::visit(const VectorReduce *op) {
+    op->value.accept(this);
+
+    // represents the number of times the op->op is applied to the vector
+    int count_cost = op->value.type().lanes() - 1;
+
+    std::function<int(int)> calculate_cc = [count_cost](int children_cost) {
+        return count_cost * (NORMAL_NODE_CC + children_cost);
+    };
+
+    std::function<int(int)> calculate_dmc = [count_cost](int children_cost) {
+        return count_cost * (NORMAL_NODE_DMC + children_cost);
+    };
+
+    // inclusive and exclusive costs are the same
+    set_costs(true, op, {op->value.get()}, calculate_cc, calculate_dmc);
```

Comment:
The pattern of calling `set_costs` twice with the same args (except for the first) happens so many times that it seems worthy of its own little helper function.